cpp/include/cudf/utilities/error.hpp

```diff
@@ -99,37 +100,80 @@ struct fatal_cuda_error : public cuda_error {
  * @brief Macro for checking (pre-)conditions that throws an exception when
  * a condition is violated.
  *
- * Example usage:
+ * Defaults to throwing `cudf::logic_error`, but a custom exception may also be
+ * specified.
  *
- * @code
- * CUDF_EXPECTS(lhs->dtype == rhs->dtype, "Column type mismatch");
- * @endcode
+ * Example usage:
+ * ```
+ * // throws cudf::logic_error
+ * CUDF_EXPECTS(p != nullptr, "Unexpected null pointer");
  *
- * @param[in] cond Expression that evaluates to true or false
- * @param[in] reason String literal description of the reason that cond is
- * expected to be true
- * @throw cudf::logic_error if the condition evaluates to false.
+ * // throws std::runtime_error
+ * CUDF_EXPECTS(p != nullptr, "Unexpected nullptr", std::runtime_error);
```

Comment:
Is there a reason to flip the argument order of this macro from the existing one in RMM (which is `RMM_EXPECTS(condition, error, message)`)? I would argue that consistency across the libraries is probably worth it.

Reply:
I didn't like the rmm order because it inserts an argument into the middle relative to the 2 argument form:
```
RMM_EXPECTS(_condition, _reason)
RMM_EXPECTS(_condition, _exception_type, _reason)  # Note that exception type comes _before_ reason
```
I don't know if there was a reason for that, @jrhemstad @harrism do you have any thoughts? If there's no particular reason I might instead prefer to modify the rmm macros for consistency with these new ones in cudf rather than try to match rmm.

Reply:
I've no problem with changing the order in RMM. I agree that this order is better.

Reply:
https://github.com/rapidsai/rmm/issues/1172

Reply:
No, the message isn't optional, the exception type is. You either do `RMM_EXPECTS(condition, message)` or you do `RMM_EXPECTS(condition, exception_type, message)` if you want a different exception than just an `rmm::logic_error`.